File: solutions/904-leaf-similar-trees/leaf-similar-trees.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function leafSimilar(root1: TreeNode | null, root2: TreeNode | null): boolean {
    
    const list1 = leafList(root1);
    const list2 = leafList(root2);

    if(list1.length !== list2.length) {
        return false;
    }

    for(let i = 0; i < list1.length; i += 1) {
        if(list1[i] !== list2[i]) {
            return false;
        }
    }

    return true;
};

function leafList(root: TreeNode | null):number[] {
    const result = [];
    function dfs(root) {
        if(root === null) {
            return;
        }

        if(root.left === null && root.right === null) {
            result.push(root.val);
        }

        dfs(root.left);
        dfs(root.right);
    }
    dfs(root);
    return result;
}

File: solutions/904-leaf-similar-trees/leaf-similar-trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        return self.dfs(root1) == self.dfs(root2)

    def dfs(self, root: TreeNode | None, result: list[int] | None = None) -> list[int]:
        if result is None:
            result = []
        if root is None:
            return result
        if root.left is None and root.right is None:
            result.append(root.val)
        self.dfs(root.left, result)
        self.dfs(root.right, result)
        return result
        